pallets/cash/src/internal/events.rs

```diff
@@ -1,182 +1,416 @@
-                // We need to make sure this is fully idempotent
+/// Incrementally perform the next step of tracking events from all the underlying chains.
+pub fn track_chain_events<T: Config>() -> Result<(), Reason> {
+    const OCW_TRACK_CHAIN_EVENTS: &[u8; 24] = b"cash::track_chain_events";
+    let mut lock = StorageLock::<Time>::new(OCW_TRACK_CHAIN_EVENTS);
+    let result = match lock.try_lock() {
+        Ok(_guard) => {
+            // Note: chains could be parallelized
+            track_chain_events_on::<T>(ChainId::Eth)
+        }
 
-                // Send extrinsics for all events
-                submit_events::<T>(event_info.events)?;
+        _ => Err(Reason::WorkerBusy),
+    };
+    result
+}
 
-                // Save latest block in ocw storage
-                s_info.set(&event_info.latest_eth_block);
-            }
-            Err(err) => {
-                log!("Error while fetching events: {:?}", err);
-                return Err(Reason::FetchError);
+/// Perform the next step of tracking events from an underlying chain.
+pub fn track_chain_events_on<T: Config>(chain_id: ChainId) -> Result<(), Reason> {
+    let me = sp_core::crypto::AccountId32::new([0u8; 32]); // XXX self worker identity
+    let last_block = LastProcessedBlock::get(chain_id).ok_or(Reason::NoSuchBlock)?;
+    let true_block = fetch_chain_block(chain_id, last_block.number())?;
+    if last_block.hash() == true_block.hash() {
+        // worker sees same fork as chain
+        let pending_blocks = PendingChainBlocks::get(chain_id);
+        let event_queue = IngressionQueue::get(chain_id).ok_or(Reason::NoSuchQueue)?;
+        let slack = queue_slack(chain_id, &event_queue);
+        let blocks = fetch_chain_blocks(chain_id, last_block.number(), slack)?;
+        submit_chain_blocks::<T>(chain_id, blocks)
+    } else {
+        // worker sees a different fork
+        let reorg = formulate_reorg::<T>(chain_id, &last_block, &true_block)?;
+        if !already_voted_for_reorg::<T>(chain_id, me, &reorg) {
+            submit_chain_reorg::<T>(chain_id, &reorg)
+        } else {
+            // just wait for the reorg to succeed or fail,
+            //  or we change our minds in another pass (noop)
+            Ok(())
+        }
+    }
+}
+
+/// Ingress a single round (quota per underlying chain block ingested).
+pub fn ingress_queue<T: Config>(
+    chain_id: ChainId,
+    last_block: &ChainBlock,
+    event_queue: &mut ChainBlockEvents,
+) -> Result<(), Reason> {
+    use crate::types::{Quantity, USDQuantity, USD};
+    const QUOTA: USDQuantity = Quantity::from_nominal("10000", USD); // XXX value? per chain?
+    let mut available = QUOTA;
+    event_queue.retain(|event| {
+        // note that we panic if we cannot compute the risk adjusted value,
+        //  there's not an obviously more reasonable thing to do right now
+        let value = risk_adjusted_value::<T>(event, last_block.number()).unwrap();
```

Comment:
I might consider returning with no action or ignoring the event versus panicing. Panics basically prevent any progress on the chain until every node hard-forks, right?

Reply:
Doing that inside the retain creates a lot of extra complexity, and there's no real reason we can't compute this though